File: forven/routers/brain.py
"""Phase 1 (P1-T12) — Brain API: memory, decisions, recall.

Backs the SvelteKit `/brain` page (P1-T14..T17) and the Settings auxiliary
picker (P1-T18). All endpoints are auth-protected by the global
`ApiKeyMiddleware`. The shape of every response is the contract — any change
must be matched in `frontend/src/lib/api/brain.ts` (P1-T13).
"""
from __future__ import annotations

import json
from typing import Any, Literal

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, Field

from forven.api_security import require_operator_access
from forven import brain_lessons as brain_lessons_mod
from forven import recall as recall_mod
from forven.brain_decisions import get_decision
from forven.brain_memory import (
    BrainMemoryTooLargeError,
    MAX_MEMORY_CHARS,
    get_memory_with_meta,
    list_history,
    set_memory,
)
from forven.db import get_db
from forven.model_routing import (
    AUXILIARY_TASK_KINDS,
    get_model_routing,
    update_model_routing,
)

router = APIRouter(prefix="/api/brain", tags=["brain"], dependencies=[Depends(require_operator_access)])


# --------------------------------------------------------------------------- #
# Overview                                                                    #
# --------------------------------------------------------------------------- #


def _row_to_dict(row: Any) -> dict[str, Any]:
    return {k: row[k] for k in row.keys()}


def _memory_flags(body: str) -> list[dict[str, str]]:
    flags: list[dict[str, str]] = []
    lower = body.lower()
    if "critical" in lower:
        flags.append(
            {
                "kind": "memory",
                "severity": "critical",
                "title": "Memory contains critical issues",
                "detail": "Brain memory includes the word critical. Review the current state before trusting autonomous work.",
            }
        )
    if any(token in lower for token in ("blocked", "broken", "failing", "failure")):
        flags.append(
            {
                "kind": "memory",
                "severity": "warning",
                "title": "Memory reports blockers or failures",
                "detail": "Brain memory mentions blocked, broken, failing, or failure conditions.",
            }
        )
    return flags


@router.get("/overview")
def get_overview_endpoint() -> dict[str, Any]:
    """Operator-oriented Brain summary: state, activity, work, and attention."""
    memory = get_memory_with_meta()
    body = str(memory.get("body") or "")

    with get_db() as conn:
        activity_rows = conn.execute(
            "SELECT id, level, source, message, data, created_at "
            "FROM activity_log "
            "WHERE lower(COALESCE(source, '')) IN ('brain', 'agent:brain') "
            "ORDER BY created_at DESC, id DESC LIMIT 20"
        ).fetchall()

        recent_task_rows = conn.execute(
            "SELECT id, display_id, agent_id, type, title, status, strategy_id, "
            "priority, created_at, started_at, completed_at, error "
            "FROM agent_tasks "
            "WHERE COALESCE(assigned_by, '') = 'brain' OR agent_id = 'brain' "
            "ORDER BY created_at DESC, id DESC LIMIT 20"
        ).fetchall()

        active_task_rows = conn.execute(
            "SELECT id, display_id, agent_id, type, title, status, strategy_id, "
            "priority, created_at, started_at, completed_at, error "
            "FROM agent_tasks "
            "WHERE (COALESCE(assigned_by, '') = 'brain' OR agent_id = 'brain') "
            "AND lower(COALESCE(status, '')) IN "
            "('pending', 'running', 'blocked', 'paused_manual', 'failed') "
            "ORDER BY "
            "CASE lower(COALESCE(status, '')) "
            "WHEN 'failed' THEN 0 WHEN 'blocked' THEN 1 WHEN 'running' THEN 2 "
            "WHEN 'pending' THEN 3 ELSE 4 END, "
            "created_at DESC, id DESC LIMIT 20"
        ).fetchall()

        active_task_count = int(
            conn.execute(
                "SELECT COUNT(*) AS c FROM agent_tasks "
                "WHERE (COALESCE(assigned_by, '') = 'brain' OR agent_id = 'brain') "
                "AND lower(COALESCE(status, '')) IN "
                "('pending', 'running', 'blocked', 'paused_manual', 'failed')"
            ).fetchone()["c"]
        )

        recent_window_sql = (
            "(datetime(created_at) >= datetime('now', '-24 hours') "
            "OR created_at >= strftime('%Y-%m-%dT%H:%M:%S+00:00', 'now', '-24 hours'))"
        )

        failed_count = int(
            conn.execute(
                "SELECT COUNT(*) AS c FROM agent_tasks "
                "WHERE (COALESCE(assigned_by, '') = 'brain' OR agent_id = 'brain') "
                "AND lower(COALESCE(status, '')) = 'failed' "
                f"AND {recent_window_sql}"
            ).fetchone()["c"]
        )
        blocked_count = int(
            conn.execute(
                "SELECT COUNT(*) AS c FROM agent_tasks "
                "WHERE (COALESCE(assigned_by, '') = 'brain' OR agent_id = 'brain') "
                "AND lower(COALESCE(status, '')) IN ('blocked', 'paused_manual') "
                f"AND {recent_window_sql}"
            ).fetchone()["c"]
        )
        pending_approval_count = int(
            conn.execute(
                "SELECT COUNT(*) AS c FROM approvals "
                "WHERE lower(COALESCE(status, '')) IN ('pending_approval', 'pending')"
            ).fetchone()["c"]
        )
        decisions_count = int(
            conn.execute("SELECT COUNT(*) AS c FROM brain_decisions").fetchone()["c"]
        )
        lessons_count = int(
            conn.execute("SELECT COUNT(*) AS c FROM brain_lessons").fetchone()["c"]
        )
        recall_count = int(
            conn.execute(
                "SELECT COUNT(*) AS c FROM agent_tasks WHERE lower(COALESCE(type, '')) = 'recall'"
            ).fetchone()["c"]
        )

        repeated_failure_rows = conn.execute(
            "SELECT type, COUNT(*) AS count FROM agent_tasks "
            "WHERE (COALESCE(assigned_by, '') = 'brain' OR agent_id = 'brain') "
            "AND lower(COALESCE(status, '')) = 'failed' "
            f"AND {recent_window_sql} "
            "GROUP BY type HAVING COUNT(*) >= 3 "
            "ORDER BY count DESC LIMIT 5"
        ).fetchall()

    attention = _memory_flags(body)
    if failed_count:
        attention.append(
            {
                "kind": "tasks",
                "severity": "warning",
                "title": f"{failed_count} Brain-assigned tasks failed in the last 24h",
                "detail": "Open the active work list or Tasks page to inspect fresh recurring failures.",
            }
        )
    if blocked_count:
        attention.append(
            {
                "kind": "tasks",
                "severity": "warning",
                "title": f"{blocked_count} Brain-assigned tasks are blocked",
                "detail": "Blocked or paused work may need operator intervention.",
            }
        )
    if pending_approval_count:
        attention.append(
            {
                "kind": "approvals",
                "severity": "warning",
                "title": f"{pending_approval_count} approvals are waiting",
                "detail": "Review the Approvals page before expecting Brain to make progress.",
            }
        )
    if decisions_count == 0:
        attention.append(
            {
                "kind": "audit",
                "severity": "info",
                "title": "Decision ledger is empty",
                "detail": "Brain activity exists, but no rows have landed in brain_decisions yet.",
            }
        )
    if lessons_count == 0:
        attention.append(
            {
                "kind": "learning",
                "severity": "info",
                "title": "No validated lessons yet",
                "detail": "The learning loop is not producing curated situation-to-lesson records yet.",
            }
        )

    return {
        "memory": memory,
        "stats": {
            "activity_count": len(activity_rows),
            "active_tasks": active_task_count,
            "recent_tasks": len(recent_task_rows),
            "failed_tasks": failed_count,
            "blocked_tasks": blocked_count,
            "pending_approvals": pending_approval_count,
            "decisions": decisions_count,
            "lessons": lessons_count,
            "recalls": recall_count,
        },
        "attention": attention[:12],
        "activity": [_row_to_dict(r) for r in activity_rows],
        "active_tasks": [_row_to_dict(r) for r in active_task_rows],
        "recent_tasks": [_row_to_dict(r) for r in recent_task_rows],
        "repeated_failures": [_row_to_dict(r) for r in repeated_failure_rows],
    }


# --------------------------------------------------------------------------- #
# Memory                                                                      #
# --------------------------------------------------------------------------- #


class _MemoryPutBody(BaseModel):
    body: str = Field(..., description="Full replacement body. Max 2000 chars.")
    mutation_type: Literal["replace"] = "replace"
    mutated_by: str | None = None


@router.get("/memory")
def get_memory_endpoint() -> dict[str, Any]:
    """Return the current Brain operational memory plus metadata."""
    return get_memory_with_meta()


@router.put("/memory")
def put_memory_endpoint(body: _MemoryPutBody) -> dict[str, Any]:
    """Full overwrite of the Brain memory body. 422 if cap exceeded."""
    try:
        set_memory(
            body.body,
            mutated_by=body.mutated_by or "operator",
            mutation_type=body.mutation_type,
        )
    except BrainMemoryTooLargeError as exc:
        raise HTTPException(
            status_code=422,
            detail={
                "error": "memory_cap_exceeded",
                "current_len": exc.current_len,
                "attempted_len": exc.attempted_len,
                "cap": exc.cap,
            },
        )
    return get_memory_with_meta()


@router.get("/memory/history")
def get_memory_history_endpoint(limit: int = Query(20, ge=1, le=200)) -> dict[str, Any]:
    """Return the most recent memory mutation rows, newest first."""
    return {"history": list_history(limit=limit), "cap": MAX_MEMORY_CHARS}


# --------------------------------------------------------------------------- #
# Decisions                                                                   #
# --------------------------------------------------------------------------- #


def _decision_row_to_dict(row: Any) -> dict[str, Any]:
    out = {k: row[k] for k in row.keys()}
    decision_blob = out.get("decision_json")
    if isinstance(decision_blob, str) and decision_blob:
        try:
            out["decision"] = json.loads(decision_blob)
        except (TypeError, ValueError):
            out["decision"] = None
    else:
        out["decision"] = None
    return out


@router.get("/decisions")
def list_decisions(
    cycle_id: str | None = None,
    action_type: str | None = None,
    strategy_id: str | None = None,
    outcome: str | None = None,
    limit: int = Query(50, ge=1, le=200),
    offset: int = Query(0, ge=0),
) -> dict[str, Any]:
    """Paginated list of brain decisions with optional filters."""
    where: list[str] = []
    params: list[Any] = []

    if cycle_id:
        where.append("d.cycle_id = ?")
        params.append(cycle_id)
    if outcome:
        where.append("d.outcome_observed = ?")
        params.append(outcome)

    join_sql = ""
    if action_type or strategy_id:
        join_sql = (
            " INNER JOIN agent_tasks t ON t.brain_decision_id = d.id"
        )
        if action_type:
            where.append("t.type = ?")
            params.append(action_type)
        if strategy_id:
            where.append("t.strategy_id = ?")
            params.append(strategy_id)

    where_sql = (" WHERE " + " AND ".join(where)) if where else ""
    distinct_sql = "DISTINCT" if join_sql else ""

    with get_db() as conn:
        count_sql = (
            f"SELECT COUNT({distinct_sql} d.id) AS c FROM brain_decisions d{join_sql}{where_sql}"
        )
        total = int(conn.execute(count_sql, params).fetchone()["c"])

        list_sql = (
            f"SELECT {distinct_sql} d.id, d.cycle_id, d.situation_summary, d.decision_json, "
            f"d.action_taken, d.outcome_observed, d.outcome_at, d.prompt_hash, d.created_at "
            f"FROM brain_decisions d{join_sql}{where_sql} "
            "ORDER BY d.created_at DESC, d.id DESC LIMIT ? OFFSET ?"
        )
        rows = conn.execute(list_sql, [*params, limit, offset]).fetchall()

    return {
        "items": [_decision_row_to_dict(r) for r in rows],
        "total": total,
        "limit": limit,
        "offset": offset,
    }


@router.get("/decisions/{decision_id}")
def get_decision_endpoint(decision_id: int) -> dict[str, Any]:
    """Full decision row + the agent_tasks rows linked back to it."""
    decision = get_decision(decision_id)
    if not decision:
        raise HTTPException(status_code=404, detail="decision not found")

    decision_blob = decision.get("decision_json")
    if isinstance(decision_blob, str) and decision_blob:
        try:
            decision["decision"] = json.loads(decision_blob)
        except (TypeError, ValueError):
            decision["decision"] = None
    else:
        decision["decision"] = None

    with get_db() as conn:
        task_rows = conn.execute(
            "SELECT id, display_id, agent_id, type, title, status, strategy_id, provider, model_id, "
            "cost_usd, created_at, completed_at "
            "FROM agent_tasks WHERE brain_decision_id = ? "
            "ORDER BY created_at ASC, id ASC",
            (int(decision_id),),
        ).fetchall()
    decision["linked_tasks"] = [{k: r[k] for k in r.keys()} for r in task_rows]
    return decision


# --------------------------------------------------------------------------- #
# Recall                                                                      #
# --------------------------------------------------------------------------- #


@router.get("/recall")
def recall_endpoint(
    q: str = Query(..., description="Free-text recall query"),
    scope: Literal["all", "decisions", "tasks"] = "all",
    limit: int = Query(5, ge=1, le=20),
) -> dict[str, Any]:
    """Hybrid FTS5 + auxiliary-LLM recall. Same shape as the Brain tool envelope."""
    try:
        result = recall_mod.recall_similar_situation(q, scope=scope, limit=limit)
    except Exception as exc:  # noqa: BLE001 — recall must never 500
        return {
            "ok": False,
            "error": "recall_failed",
            "detail": str(exc),
            "summary": "",
            "hits": [],
            "aux_model": None,
            "latency_ms": 0,
        }
    return {
        "ok": True,
        "query": q,
        "scope": scope,
        "limit": limit,
        **result,
    }


# --------------------------------------------------------------------------- #
# Auxiliary model routing (Settings -> Models)                                #
# --------------------------------------------------------------------------- #


class _AuxiliaryEntry(BaseModel):
    # provider/model_id are Optional so the frontend can send
    # ``{"provider": null, "model_id": null}`` (or empty strings) to mean
    # "remove this slot". A non-empty entry is a normal merge; an empty one is
    # treated as a DELETE in put_auxiliary_endpoint.
    provider: str | None = None
    model_id: str | None = None
    base_url: str | None = None
    api_key: str | None = None


class _AuxiliaryUpdateBody(BaseModel):
    auxiliary: dict[str, _AuxiliaryEntry]


@router.get("/auxiliary")
def get_auxiliary_endpoint() -> dict[str, Any]:
    """Return the four auxiliary slots (compression / recall / skill_extraction / post_mortem)."""
    policy = get_model_routing()
    aux = policy.get("auxiliary") or {}
    return {
        "auxiliary": {
            kind: aux.get(kind, {"provider": None, "model_id": None, "base_url": None, "api_key": None})
            for kind in AUXILIARY_TASK_KINDS
        },
        "task_kinds": list(AUXILIARY_TASK_KINDS),
    }


@router.put("/auxiliary")
def put_auxiliary_endpoint(body: _AuxiliaryUpdateBody) -> dict[str, Any]:
    """Replace one or more auxiliary slots. Only known task kinds are accepted."""
    invalid = [k for k in body.auxiliary.keys() if k not in AUXILIARY_TASK_KINDS]
    if invalid:
        raise HTTPException(
            status_code=422,
            detail={"error": "unknown_task_kind", "unknown": invalid, "valid": list(AUXILIARY_TASK_KINDS)},
        )
    current = get_model_routing()
    next_aux = dict(current.get("auxiliary") or {})
    for kind, entry in body.auxiliary.items():
        provider = (entry.provider or "").strip()
        model_id = (entry.model_id or "").strip()
        if not provider or not model_id:
            # Cleared slot: an entry missing provider OR model_id means
            # "remove this slot". Pop it so it is not merged back in (and so the
            # coercion in update_model_routing re-seeds nothing for it).
            next_aux.pop(kind, None)
            continue
        next_aux[kind] = {
            "provider": provider,
            "model_id": model_id,
            "base_url": entry.base_url,
            "api_key": entry.api_key,
        }
    next_policy = {**current, "auxiliary": next_aux}
    update_model_routing(next_policy)
    return get_auxiliary_endpoint()


# --------------------------------------------------------------------------- #
# Brain lessons (P3-T08)                                                      #
# --------------------------------------------------------------------------- #


class _LessonCreateBody(BaseModel):
    situation_pattern: str
    lesson_text: str
    evidence_decisions: list[int] = Field(default_factory=list)
    confidence: float = 0.5


class _LessonUpdateBody(BaseModel):
    situation_pattern: str | None = None
    lesson_text: str | None = None
    confidence: float | None = None
    last_validated_at: str | None = None


@router.get("/lessons")
def list_lessons_endpoint(
    limit: int = Query(50, ge=1, le=200),
    min_confidence: float = Query(0.0, ge=0.0, le=1.0),
) -> dict[str, Any]:
    """Paginated list of brain lessons. Filter by minimum confidence."""
    items = brain_lessons_mod.list_lessons(limit=limit, min_confidence=min_confidence)
    return {"items": items, "count": len(items)}


@router.get("/lessons/search")
def search_lessons_endpoint(
    q: str = Query(..., description="FTS5 query over situation_pattern + lesson_text"),
    limit: int = Query(20, ge=1, le=50),
) -> dict[str, Any]:
    items = brain_lessons_mod.search_lessons(q, limit=limit)
    return {"query": q, "items": items, "count": len(items)}


@router.post("/lessons")
def create_lesson_endpoint(body: _LessonCreateBody) -> dict[str, Any]:
    try:
        return brain_lessons_mod.create_lesson(
            situation_pattern=body.situation_pattern,
            lesson_text=body.lesson_text,
            evidence_decisions=body.evidence_decisions,
            confidence=body.confidence,
            created_by="operator",
        )
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc


@router.get("/lessons/{lesson_id}")
def get_lesson_endpoint(lesson_id: int) -> dict[str, Any]:
    row = brain_lessons_mod.get_lesson(lesson_id)
    if row is None:
        raise HTTPException(status_code=404, detail=f"lesson {lesson_id} not found")
    return row


@router.put("/lessons/{lesson_id}")
def update_lesson_endpoint(lesson_id: int, body: _LessonUpdateBody) -> dict[str, Any]:
    try:
        row = brain_lessons_mod.update_lesson(
            lesson_id,
            situation_pattern=body.situation_pattern,
            lesson_text=body.lesson_text,
            confidence=body.confidence,
            last_validated_at=body.last_validated_at,
        )
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    if row is None:
        raise HTTPException(status_code=404, detail=f"lesson {lesson_id} not found")
    return row


@router.delete("/lessons/{lesson_id}")
def delete_lesson_endpoint(lesson_id: int) -> dict[str, Any]:
    ok = brain_lessons_mod.delete_lesson(lesson_id)
    if not ok:
        raise HTTPException(status_code=404, detail=f"lesson {lesson_id} not found")
    return {"ok": True, "lesson_id": lesson_id}


@router.post("/lessons/{lesson_id}/validate")
def validate_lesson_endpoint(lesson_id: int) -> dict[str, Any]:
    """Stamp `last_validated_at` to now (UTC). Convenience for the Lessons tab."""
    row = brain_lessons_mod.mark_validated(lesson_id)
    if row is None:
        raise HTTPException(status_code=404, detail=f"lesson {lesson_id} not found")
    return row


__all__ = ["router"]

File: frontend/src/lib/api/brain.ts
/**
 * Brain API client (P1-T13) — backs the /brain page (P1-T14..T17),
 * the Settings auxiliary picker (P1-T18), and the strategy decision
 * widget (P1-T19).
 *
 * Contract is fixed by ``axiom/routers/brain.py``. Keep these
 * interfaces in lockstep with the backend payload shapes.
 */

import { fetchApi } from './core';

// --------------------------------------------------------------------------- //
// Overview                                                                    //
// --------------------------------------------------------------------------- //

export interface BrainOverviewMemory {
	body: string;
	updated_at: string | null;
	updated_by: string | null;
	char_count: number;
	cap: number;
}

export interface BrainOverviewStats {
	activity_count: number;
	active_tasks: number;
	recent_tasks: number;
	failed_tasks: number;
	blocked_tasks: number;
	pending_approvals: number;
	decisions: number;
	lessons: number;
	recalls: number;
}

export interface BrainAttentionItem {
	kind: string;
	severity: 'critical' | 'warning' | 'info' | string;
	title: string;
	detail: string;
}

export interface BrainActivityRow {
	id: number;
	level: string;
	source: string | null;
	message: string;
	data: string | null;
	created_at: string;
}

export interface BrainOverviewTask {
	id: number;
	display_id: string | null;
	agent_id: string | null;
	type: string | null;
	title: string | null;
	status: string | null;
	strategy_id: string | null;
	priority: number | null;
	created_at: string | null;
	started_at: string | null;
	completed_at: string | null;
	error: string | null;
}

export interface BrainRepeatedFailure {
	type: string | null;
	count: number;
}

export interface BrainOverview {
	memory: BrainOverviewMemory;
	stats: BrainOverviewStats;
	attention: BrainAttentionItem[];
	activity: BrainActivityRow[];
	active_tasks: BrainOverviewTask[];
	recent_tasks: BrainOverviewTask[];
	repeated_failures: BrainRepeatedFailure[];
}

export async function getBrainOverview(): Promise<BrainOverview> {
	return fetchApi<BrainOverview>('/brain/overview');
}

// --------------------------------------------------------------------------- //
// Memory                                                                      //
// --------------------------------------------------------------------------- //

export interface BrainMemoryState {
	body: string;
	updated_at: string | null;
	updated_by: string | null;
	char_count: number;
	cap: number;
}

export type BrainMemoryMutationType = 'replace' | 'add' | 'remove';

export interface BrainMemoryHistoryRow {
	id: number;
	mutation_type: BrainMemoryMutationType;
	before_excerpt: string | null;
	after_excerpt: string | null;
	mutated_at: string;
	mutated_by: string | null;
}

export interface BrainMemoryHistoryResponse {
	history: BrainMemoryHistoryRow[];
	cap: number;
}

export interface BrainMemoryCapErrorDetail {
	error: 'memory_cap_exceeded';
	current_len: number;
	attempted_len: number;
	cap: number;
}

export async function getBrainMemory(): Promise<BrainMemoryState> {
	return fetchApi<BrainMemoryState>('/brain/memory');
}

export async function putBrainMemory(
	body: string,
	options?: { mutatedBy?: string }
): Promise<BrainMemoryState> {
	return fetchApi<BrainMemoryState>('/brain/memory', {
		method: 'PUT',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({
			body,
			mutation_type: 'replace',
			mutated_by: options?.mutatedBy ?? 'operator'
		})
	});
}

export async function getBrainMemoryHistory(
	limit = 20
): Promise<BrainMemoryHistoryResponse> {
	const params = new URLSearchParams({ limit: String(limit) });
	return fetchApi<BrainMemoryHistoryResponse>(
		`/brain/memory/history?${params.toString()}`
	);
}

// --------------------------------------------------------------------------- //
// Decisions                                                                   //
// --------------------------------------------------------------------------- //

export type BrainDecisionOutcome = 'success' | 'failure' | 'mixed' | string;

export interface BrainDecisionRow {
	id: number;
	cycle_id: string | null;
	situation_summary: string | null;
	decision_json: string | null;
	decision: unknown | null;
	action_taken: string | null;
	outcome_observed: BrainDecisionOutcome | null;
	outcome_at: string | null;
	prompt_hash: string | null;
	created_at: string;
}

export interface BrainDecisionsListResponse {
	items: BrainDecisionRow[];
	total: number;
	limit: number;
	offset: number;
}

export interface BrainDecisionLinkedTask {
	id: number;
	display_id: string | null;
	agent_id: string | null;
	type: string | null;
	title: string | null;
	status: string | null;
	strategy_id: string | null;
	provider: string | null;
	model_id: string | null;
	cost_usd: number | null;
	created_at: string | null;
	completed_at: string | null;
}

export interface BrainDecisionDetail extends BrainDecisionRow {
	linked_tasks: BrainDecisionLinkedTask[];
}

export interface BrainDecisionsListQuery {
	cycleId?: string;
	actionType?: string;
	strategyId?: string;
	outcome?: BrainDecisionOutcome;
	limit?: number;
	offset?: number;
}

export async function getBrainDecisions(
	query: BrainDecisionsListQuery = {}
): Promise<BrainDecisionsListResponse> {
	const params = new URLSearchParams();
	if (query.cycleId) params.set('cycle_id', query.cycleId);
	if (query.actionType) params.set('action_type', query.actionType);
	if (query.strategyId) params.set('strategy_id', query.strategyId);
	if (query.outcome) params.set('outcome', query.outcome);
	if (query.limit !== undefined) params.set('limit', String(query.limit));
	if (query.offset !== undefined) params.set('offset', String(query.offset));
	const qs = params.toString();
	return fetchApi<BrainDecisionsListResponse>(
		qs ? `/brain/decisions?${qs}` : '/brain/decisions'
	);
}

export async function getBrainDecision(
	decisionId: number
): Promise<BrainDecisionDetail> {
	return fetchApi<BrainDecisionDetail>(`/brain/decisions/${decisionId}`);
}

// --------------------------------------------------------------------------- //
// Recall                                                                      //
// --------------------------------------------------------------------------- //

export type BrainRecallScope = 'all' | 'decisions' | 'tasks';
export type BrainRecallSource = 'brain_decisions' | 'agent_tasks';

export interface BrainRecallHit {
	source: BrainRecallSource;
	id: number;
	score: number;
	rerank_score?: number;
	snippet: string | null;
	situation: string | null;
	outcome: string | null;
	created_at: string | null;
	deep_link_url: string;
}

export interface BrainRecallSuccess {
	ok: true;
	query: string;
	scope: BrainRecallScope;
	limit: number;
	summary: string;
	hits: BrainRecallHit[];
	aux_model: string | null;
	latency_ms: number;
}

export interface BrainRecallFailure {
	ok: false;
	error: 'recall_failed' | string;
	detail?: string;
	summary: '';
	hits: [];
	aux_model: null;
	latency_ms: 0;
}

export type BrainRecallResponse = BrainRecallSuccess | BrainRecallFailure;

export async function recallSimilarSituation(
	q: string,
	options: { scope?: BrainRecallScope; limit?: number } = {}
): Promise<BrainRecallResponse> {
	const params = new URLSearchParams({ q });
	if (options.scope) params.set('scope', options.scope);
	if (options.limit !== undefined) params.set('limit', String(options.limit));
	return fetchApi<BrainRecallResponse>(`/brain/recall?${params.toString()}`);
}

// --------------------------------------------------------------------------- //
// Auxiliary model routing                                                     //
// --------------------------------------------------------------------------- //

export type BrainAuxiliaryTaskKind =
	| 'compression'
	| 'recall'
	| 'skill_extraction'
	| 'post_mortem';

export interface BrainAuxiliaryEntry {
	provider: string | null;
	model_id: string | null;
	base_url: string | null;
	api_key: string | null;
}

export interface BrainAuxiliaryResponse {
	auxiliary: Record<BrainAuxiliaryTaskKind, BrainAuxiliaryEntry>;
	task_kinds: BrainAuxiliaryTaskKind[];
}

export async function getBrainAuxiliary(): Promise<BrainAuxiliaryResponse> {
	return fetchApi<BrainAuxiliaryResponse>('/brain/auxiliary');
}

export async function updateBrainAuxiliary(
	auxiliary: Partial<Record<BrainAuxiliaryTaskKind, BrainAuxiliaryEntry>>
): Promise<BrainAuxiliaryResponse> {
	return fetchApi<BrainAuxiliaryResponse>('/brain/auxiliary', {
		method: 'PUT',
		body: JSON.stringify({ auxiliary })
	});
}

// --------------------------------------------------------------------------- //
// Brain lessons (P3-T10) — backs the /brain Lessons tab (P3-T13).             //
// --------------------------------------------------------------------------- //

export interface BrainLesson {
	id: number;
	situation_pattern: string;
	lesson_text: string;
	evidence_decisions: number[];
	confidence: number;
	created_at: string;
	created_by: string | null;
	last_validated_at: string | null;
}

export interface BrainLessonsListResponse {
	items: BrainLesson[];
	count: number;
}

export interface BrainLessonsSearchResponse extends BrainLessonsListResponse {
	query: string;
}

export interface BrainLessonCreateBody {
	situation_pattern: string;
	lesson_text: string;
	evidence_decisions?: number[];
	confidence?: number;
}

export interface BrainLessonUpdateBody {
	situation_pattern?: string;
	lesson_text?: string;
	confidence?: number;
	last_validated_at?: string | null;
}

export async function listBrainLessons(
	options: { limit?: number; minConfidence?: number } = {}
): Promise<BrainLessonsListResponse> {
	const params = new URLSearchParams();
	if (options.limit !== undefined) params.set('limit', String(options.limit));
	if (options.minConfidence !== undefined)
		params.set('min_confidence', String(options.minConfidence));
	const qs = params.toString();
	return fetchApi<BrainLessonsListResponse>(qs ? `/brain/lessons?${qs}` : '/brain/lessons');
}

export async function searchBrainLessons(
	query: string,
	limit = 20
): Promise<BrainLessonsSearchResponse> {
	const params = new URLSearchParams({ q: query, limit: String(limit) });
	return fetchApi<BrainLessonsSearchResponse>(`/brain/lessons/search?${params.toString()}`);
}

export async function createBrainLesson(body: BrainLessonCreateBody): Promise<BrainLesson> {
	return fetchApi<BrainLesson>('/brain/lessons', {
		method: 'POST',
		body: JSON.stringify({
			evidence_decisions: [],
			confidence: 0.5,
			...body
		})
	});
}

export async function getBrainLesson(lessonId: number): Promise<BrainLesson> {
	return fetchApi<BrainLesson>(`/brain/lessons/${lessonId}`);
}

export async function updateBrainLesson(
	lessonId: number,
	body: BrainLessonUpdateBody
): Promise<BrainLesson> {
	return fetchApi<BrainLesson>(`/brain/lessons/${lessonId}`, {
		method: 'PUT',
		body: JSON.stringify(body)
	});
}

export async function deleteBrainLesson(
	lessonId: number
): Promise<{ ok: true; lesson_id: number }> {
	return fetchApi(`/brain/lessons/${lessonId}`, { method: 'DELETE' });
}

export async function validateBrainLesson(lessonId: number): Promise<BrainLesson> {
	return fetchApi<BrainLesson>(`/brain/lessons/${lessonId}/validate`, { method: 'POST' });
}
